### PR TITLE
TST Adds test for non-contiguous tensors

### PR DESCRIPTION
**Pull request by thomasjpfan**
Follow up to #61935

This PR:

1. Adds test for non-contiguous tensors
2. Fixes bug in `NLLLoss` that was catch by the test.

The reason this was not catch in `common_nn` is because `CriterionTest` overrides `test_cuda` but does not call `test_nonconfig`.

cc @albanD @mruberry @jbschlosser @walterddr

### CODE REVIEW

File: test/test_modules.py [hunk 1]
```diff
@@ -206,6 +207,121 @@ def test_check_inplace(self, device, dtype, module_info):
...
+            elif isinstance(item, torch.Tensor) and item.requires_grad:
+                item.retain_grad()
+
+        def _get_grads(obj):
+            if isinstance(obj, tuple):
+                return tuple(_get_grads(o) for o in obj)
+            elif isinstance(obj, dict):
+                return {name: _get_grads(o) for name, o in obj.items()}
+            elif isinstance(obj, torch.Tensor) and obj.requires_grad:
+                return obj.grad
```
Comment (thomasjpfan) [inline test/test_modules.py:236]: Maybe these should be methods to the test class since they are so common to do in gradient checks.

File: test/test_modules.py [hunk 2]
```diff
@@ -206,6 +207,121 @@ def test_check_inplace(self, device, dtype, module_info):
...
+        def _zero_grad(obj):
+            if isinstance(obj, (tuple, list)) or isgenerator(obj):
+                for o in obj:
+                    _zero_grad(o)
+            elif isinstance(obj, dict):
+                for o in obj.values():
+                    _zero_grad(o)
+            elif isinstance(obj, (torch.Tensor, torch.nn.Parameter)) and obj.grad is not None:
+                obj.grad.detach_()
+                obj.grad.zero_()
```
Comment (thomasjpfan) [inline test/test_modules.py:247]: When we start to combine tests together, this will be very common operation to reset the gradients.

File: test/test_modules.py [hunk 3]
```diff
@@ -205,6 +206,120 @@ def test_check_inplace(self, device, dtype, module_info):
...
+    def _zero_grad(self, obj):
+        if isinstance(obj, (tuple, list)) or isgenerator(obj):
+            for o in obj:
+                self._zero_grad(o)
+        elif isinstance(obj, dict):
+            for o in obj.values():
+                self._zero_grad(o)
+        elif isinstance(obj, (torch.Tensor, torch.nn.Parameter)) and obj.grad is not None:
+            obj.grad.detach_()
+            obj.grad.zero_()
```
Comment (jbschlosser) [inline test/test_modules.py:238]: Hm I'm starting to see some patterns here - I wonder if there's some more generic way we can do arbitrary things over all inputs?

**Review by jbschlosser — COMMENTED** [review]
Looking pretty good! Couple small things below

File: test/test_modules.py [hunk 4]
```diff
@@ -205,6 +206,115 @@ def test_check_inplace(self, device, dtype, module_info):
...
+            if obj.requires_grad:
+                return obj.grad
+        return self._traverse_obj(obj, inner_get_grad)
+
+    def _zero_grad(self, obj):
+        def inner_zero_grad(obj):
+            if obj.grad is not None:
+                obj.grad.detach_()
+                obj.grad.zero_()
+        self._traverse_obj(obj, inner_zero_grad)
```
Comment (jbschlosser) [inline test/test_modules.py:236]: Cool, I like the factored out version :) Should these be moved to utils?

Comment (thomasjpfan) [reply test/test_modules.py]: I would keep them here and refactor them into `utils` if there are other places in the codebase that have the same pattern?

**Review by jbschlosser — APPROVED** [review]
LGTM!
